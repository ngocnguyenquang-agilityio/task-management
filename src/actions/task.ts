'use server';

import { redirect } from 'next/navigation';
import { revalidateTag } from 'next/cache';

// Constants
import {
  COLLECTION,
  ERROR_MESSAGES,
  ROUTES,
  TAGS,
  TASK_PRIORITY_VALUE,
  TASK_STATUS_VALUE,
  TaskFormDataSchema,
} from '@/constants';

// DB
import { addDocument } from '@/db';

// Models
import { Task, TaskFormState, TaskFormType } from '@/models';

// HOCs
import { withAuth } from '@/hocs';

export const createTask = async (
  prevState: TaskFormState,
  values: TaskFormType,
) => {
  return await withAuth<
    {
      prevState: TaskFormState;
      values: TaskFormType;
    },
    TaskFormState
  >(
    async (args, session) => {
      const { values } = args;

      const validators = TaskFormDataSchema.safeParse(values);

      let result: TaskFormState = {};

      if (validators.success && session?.user.id) {
        result = { success: true };

        try {
          const time = new Date().toISOString();

          const data: Omit<Task, 'id'> = {
            slug: values.slug,
            title: values.title,
            description: values.description,
            assignedTo: values.assignedTo,
            createdAt: time,
            createdBy: session.user.id,
            dueDate: values.dueDate,
            image: values.image,
            isArchived: false,
            priority: values.priority as TASK_PRIORITY_VALUE,
            status: values.status as TASK_STATUS_VALUE,
            updatedAt: time,
            projectId: values.projectId,
          };

          const taskResponse = await addDocument<Omit<Task, 'id'>>(
            COLLECTION.TASKS,
            data,
          );

          if (!taskResponse?.success) {
            throw new Error(ERROR_MESSAGES.UPSERTING_DATA_ERROR('Task'));
          }
        } catch (error) {
          result = {
            success: false,
            error: (error as Error).message,
          };
        }
      }

      if (validators.error) {
        result = {
          success: false,
          errors: validators.error.flatten().fieldErrors,
        };
      }

      if (result.success) {
        revalidateTag(TAGS.TASK_LIST);
        redirect(ROUTES.ADMIN_TASK_LIST);
      }

      return result;
    },
    { prevState, values },
  );
};
