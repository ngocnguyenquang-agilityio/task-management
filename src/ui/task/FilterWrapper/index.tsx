'use client';

import { useCallback, useMemo, useTransition } from 'react';
import { useSearchParams, useRouter } from 'next/navigation';

// Components
import { Dropdown, MultipleSelect } from '@/components';

// Types
import { OptionType } from '@/types';

// Constants
import { SEARCH_PARAMS } from '@/constants';

// Utils
import { getQueryParams } from '@/utils';

export const STATUS_OPTIONS: OptionType[] = [
  {
    name: 'Not Started',
    value: 'not_started',
  },
  {
    name: 'In Progress',
    value: 'in_progress',
  },
  {
    name: 'Done',
    value: 'done',
  },
];

export const PRIORITY_OPTIONS: OptionType[] = [
  {
    name: 'Low',
    value: 'low',
  },
  {
    name: 'Medium',
    value: 'medium',
  },
  {
    name: 'High',
    value: 'high',
  },
];

export const SORT_OPTIONS: OptionType[] = [
  {
    name: 'Desc',
    value: 'desc',
  },
  {
    name: 'Asc',
    value: 'asc',
  },
];

export const FilterWrapper = () => {
  const router = useRouter();
  const searchParams = useSearchParams();

  const [_, startTransition] = useTransition();

  const params = useMemo(
    () => new URLSearchParams(searchParams),
    [searchParams],
  );

  // store the current query parameters
  const page = useMemo(() => searchParams.get('page') || '', [searchParams]);
  const sortBy = useMemo(
    () => searchParams.get('sortBy') || '',
    [searchParams],
  );
  const status = useMemo(
    () => searchParams.get('status') || '',
    [searchParams],
  );
  const priority = useMemo(
    () => searchParams.get('priority') || '',
    [searchParams],
  );

  const handleStatusOnChange = useCallback(
    (listStatus: string[]) => {
      const status = encodeURIComponent(listStatus.join(','));

      params.set(SEARCH_PARAMS.STATUS, status);
      startTransition(() => {
        router.push(getQueryParams({ page, sortBy, status, priority }));
      });
    },
    [page, params, priority, router, sortBy],
  );

  const handlePriorityOnChange = useCallback(
    (listPriority: string[]) => {
      const priority = encodeURIComponent(listPriority.join(','));

      params.set(SEARCH_PARAMS.PRIORITY, priority);
      startTransition(() => {
        router.push(getQueryParams({ page, sortBy, priority, status }));
      });
    },
    [page, params, router, sortBy, status],
  );

  const handleSelectSort = useCallback(
    (sortBy: string) => {
      params.set(SEARCH_PARAMS.SORT_BY, sortBy);

      startTransition(() => {
        router.push(getQueryParams({ page, sortBy, status, priority }));
      });
    },
    [params, router, page, status, priority],
  );

  return (
    <div className="mb-6">
      <div className="flex gap-2.5">
        <div className="flex flex-col gap-2.5 md:flex-row">
          <MultipleSelect
            title="Status"
            onChange={handleStatusOnChange}
            options={STATUS_OPTIONS}
          />
          <MultipleSelect
            title="Priority"
            onChange={handlePriorityOnChange}
            options={PRIORITY_OPTIONS}
          />
        </div>
        <Dropdown
          placeholder="Sort"
          customClass={{
            placeholder: 'text-black',
            button: 'pb-[11px] pt-3 px-5',
          }}
          options={SORT_OPTIONS}
          onSelect={handleSelectSort}
        />
      </div>
    </div>
  );
};
