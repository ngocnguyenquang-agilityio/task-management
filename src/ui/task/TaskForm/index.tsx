'use client';

import { useEffect, useMemo } from 'react';
import { useFormStatus } from 'react-dom';
import { Control, Controller, UseFormSetValue, useForm } from 'react-hook-form';
import { zodResolver } from '@hookform/resolvers/zod';
import dayjs from 'dayjs';

// Components
import { Button, Dropdown, Input, Text } from '@/components';

// Constants
import {
  DATE_FORMAT,
  TASK_PRIORITY_OPTIONS,
  TASK_STATUS_OPTIONS,
  TaskFormDataSchema,
} from '@/constants';

// Models
import { User } from '@/types';
import { Task, Project, TaskFormState, TaskFormType } from '@/models';

// Utils
import {
  cn,
  isEmpty,
  isEnableSubmitButton,
  formatDate,
  setServerActionErrors,
  generateSlug,
} from '@/utils';

const DEFAULT_REQUIRED_FIELDS = [
  'title',
  'description',
  'status',
  'priority',
  'assignedTo',
  'projectId',
];

type TaskFormProps = {
  assginedToOptions: User[];
  fromProject: Project[];
  taskValue?: Task;
  state: TaskFormState;
  onSubmit: (formValues: TaskFormType) => void;
};

type TaskFormContentType = {
  assginedToOptions: User[];
  listProject: Project[];
  control: Control<{
    title: string;
    slug: string;
    description: string;
    image?: string;
    status: string;
    priority: string;
    assignedTo: string;
    projectId: string;
    dueDate: Date;
  }>;
  setValue: UseFormSetValue<{
    title: string;
    status: string;
    slug: string;
    description: string;
    dueDate: Date;
    priority: string;
    assignedTo: string;
    projectId: string;
    image?: string | undefined;
  }>;
  responseMessage?: string;
  isDisabled: boolean;
  isCreated?: boolean;
};

const TaskFormContent = ({
  assginedToOptions,
  listProject,
  control,
  setValue,
  responseMessage,
  isDisabled,
  isCreated,
}: TaskFormContentType) => {
  const { pending } = useFormStatus();

  return (
    <div className="">
      <Controller
        name="title"
        control={control}
        render={({
          field: { onChange, value, ...rest },
          fieldState: { error },
        }) => (
          <div className="flex flex-col gap-2">
            <label className="font-bold text-md">Title</label>
            <Input
              placeholder="Title"
              value={value}
              onChange={(value) => {
                onChange(value);
                setValue('slug', generateSlug(value.target.value));
              }}
              customClass="py-5"
              {...rest}
            />
            <span className={cn('bg-white', error?.message ? 'mb-2' : 'mb-8')}>
              {error?.message && (
                <Text
                  customClass="text-xs px-0 whitespace-pre"
                  value={error?.message}
                  variant="error"
                />
              )}
            </span>
          </div>
        )}
      />
      <Controller
        name="slug"
        control={control}
        render={({
          field: { onChange, value, ...rest },
          fieldState: { error },
        }) => (
          <div className="flex flex-col gap-2">
            <label className="font-bold text-md">Slug</label>
            <Input
              placeholder="Slug"
              value={value}
              onChange={(value) => {
                onChange(value);
              }}
              customClass="py-5"
              {...rest}
            />
            <span className={cn('bg-white', error?.message ? 'mb-2' : 'mb-8')}>
              {error?.message && (
                <Text
                  customClass="text-xs px-0 whitespace-pre"
                  value={error?.message}
                  variant="error"
                />
              )}
            </span>
          </div>
        )}
      />
      <Controller
        name="description"
        control={control}
        render={({
          field: { onChange, value, ...rest },
          fieldState: { error },
        }) => (
          <div className="flex flex-col gap-2">
            <label className="font-bold text-md">Description</label>
            <Input
              placeholder="Description"
              value={value}
              onChange={(value) => {
                onChange(value);
              }}
              customClass="py-5"
              {...rest}
            />
            <span className={cn('bg-white', error?.message ? 'mb-2' : 'mb-8')}>
              {error?.message && (
                <Text
                  customClass="text-xs px-0 whitespace-pre"
                  value={error?.message}
                  variant="error"
                />
              )}
            </span>
          </div>
        )}
      />
      <Controller
        name="image"
        control={control}
        render={({
          field: { onChange, value, ...rest },
          fieldState: { error },
        }) => (
          <div className="flex flex-col gap-2">
            <label className="font-bold text-md">Image</label>
            <Input
              placeholder="Image"
              value={value}
              onChange={(value) => {
                onChange(value);
              }}
              customClass="py-5"
              {...rest}
            />
            <span className={cn('bg-white', error?.message ? 'mb-2' : 'mb-8')}>
              {error?.message && (
                <Text
                  customClass="text-xs px-0 whitespace-pre"
                  value={error?.message}
                  variant="error"
                />
              )}
            </span>
          </div>
        )}
      />
      <div className="flex flex-col sm:flex-row sm:gap-4">
        <Controller
          name="status"
          control={control}
          render={({
            field: { onChange, value, onBlur },
            fieldState: { error },
          }) => (
            <div className="flex flex-col gap-2 basis-1/2">
              <label className="font-bold text-md">Status</label>
              <Dropdown
                placeholder="Status"
                options={TASK_STATUS_OPTIONS}
                selectedItemValue={value}
                onSelect={(value) => {
                  onChange(value);
                }}
                onBlur={onBlur}
              />
              <span
                className={cn('bg-white', error?.message ? 'mb-2' : 'mb-8')}
              >
                {error?.message && (
                  <Text
                    customClass="text-xs px-0 whitespace-pre"
                    value={error?.message}
                    variant="error"
                  />
                )}
              </span>
            </div>
          )}
        />
        <Controller
          name="priority"
          control={control}
          render={({
            field: { onChange, value, onBlur },
            fieldState: { error },
          }) => (
            <div className="flex flex-col gap-2 basis-1/2 z-10">
              <label className="font-bold text-md">Priority</label>
              <Dropdown
                placeholder="Priority"
                options={TASK_PRIORITY_OPTIONS}
                selectedItemValue={value}
                onSelect={(value) => {
                  onChange(value);
                }}
                onBlur={onBlur}
              />
              <span
                className={cn('bg-white', error?.message ? 'mb-2' : 'mb-8')}
              >
                {error?.message && (
                  <Text
                    customClass="text-xs px-0 whitespace-pre"
                    value={error?.message}
                    variant="error"
                  />
                )}
              </span>
            </div>
          )}
        />
      </div>

      <Controller
        name="assignedTo"
        control={control}
        render={({
          field: { onChange, value, onBlur },
          fieldState: { error },
        }) => (
          <div className="flex flex-col gap-2">
            <label className="font-bold text-md">Assigned To</label>
            <Dropdown
              placeholder="Assigned To User"
              options={assginedToOptions.map((user) => ({
                name: user.name,
                value: user.id,
              }))}
              selectedItemValue={value}
              onSelect={(value) => {
                onChange(value);
              }}
              onBlur={onBlur}
            />
            <span className={cn('bg-white', error?.message ? 'mb-2' : 'mb-8')}>
              {error?.message && (
                <Text
                  customClass="text-xs px-0 whitespace-pre"
                  value={error?.message}
                  variant="error"
                />
              )}
            </span>
          </div>
        )}
      />

      <Controller
        name="projectId"
        control={control}
        render={({
          field: { onChange, value, onBlur },
          fieldState: { error },
        }) => (
          <div className="flex flex-col gap-2">
            <label className="font-bold text-md">Project</label>
            <Dropdown
              placeholder="Select a Project"
              options={listProject.map((project) => ({
                name: project.title,
                value: project.id,
              }))}
              selectedItemValue={value}
              onSelect={(value) => {
                onChange(value);
              }}
              onBlur={onBlur}
            />
            <span className={cn('bg-white', error?.message ? 'mb-2' : 'mb-8')}>
              {error?.message && (
                <Text
                  customClass="text-xs px-0 whitespace-pre"
                  value={error?.message}
                  variant="error"
                />
              )}
            </span>
          </div>
        )}
      />
      <Controller
        name="dueDate"
        control={control}
        render={({
          field: { onChange, value, ...rest },
          fieldState: { error },
        }) => (
          <div className="flex flex-col gap-2">
            <label className="font-bold text-md">Due Date</label>
            <Input
              placeholder="Due Date"
              type="date"
              value={formatDate(value, DATE_FORMAT.Tertiary)}
              onChange={(e) => {
                const date = dayjs(new Date(e.target.value));
                onChange(date);
              }}
              customClass="py-5"
              {...rest}
            />
            <span className={cn('bg-white', error?.message ? 'mb-2' : 'mb-8')}>
              {error?.message && (
                <Text
                  customClass="text-xs px-0 whitespace-pre"
                  value={error?.message}
                  variant="error"
                />
              )}
            </span>
          </div>
        )}
      />
      <Button
        type="submit"
        customClass="w-full justify-center py-[19px] font-bold mb-8"
        disabled={isDisabled}
        isLoading={pending}
      >
        {isCreated ? 'Create' : 'Edit'} Task
      </Button>
      <span className={cn('mt-3', responseMessage ? 'mb-2' : 'mb-8')}>
        {responseMessage && (
          <Text
            customClass="text-xs px-0 whitespace-pre"
            value={responseMessage}
            variant="error"
          />
        )}
      </span>
    </div>
  );
};

export const TaskForm = ({
  assginedToOptions,
  fromProject,
  taskValue,
  state,
  onSubmit,
}: TaskFormProps) => {
  const {
    title,
    slug,
    description,
    image,
    status,
    priority,
    assignedTo,
    projectId,
    dueDate,
  } = taskValue || {};

  const taskFormInitValues: TaskFormType = useMemo(
    () => ({
      title: title || '',
      slug: slug || '',
      description: description || '',
      image: image || '',
      status: status || '',
      priority: priority || '',
      assignedTo: assignedTo || '',
      projectId: projectId || '',
      dueDate: dueDate || new Date(),
    }),
    [
      title,
      slug,
      description,
      image,
      dueDate,
      status,
      priority,
      assignedTo,
      projectId,
    ],
  );

  const {
    control,
    setError,
    getValues,
    reset,
    setValue,
    formState: { dirtyFields, errors },
  } = useForm<TaskFormType>({
    mode: 'onBlur',
    reValidateMode: 'onBlur',
    values: taskFormInitValues,
    resolver: zodResolver(TaskFormDataSchema),
  });

  const dirtyItems = Object.keys(dirtyFields);
  // If create -> required fields
  // If edit -> Data is already filled ->  no empty fields
  const updateRequiredFields = useMemo(
    () => (isEmpty(taskValue) ? DEFAULT_REQUIRED_FIELDS : []),
    [taskValue],
  );
  const isDisabled = useMemo(
    () => !isEnableSubmitButton(updateRequiredFields, dirtyItems, errors),
    [dirtyItems, errors, updateRequiredFields],
  );

  const handleSubmit = () => {
    const formValues = getValues();
    onSubmit(formValues);
  };

  useEffect(() => {
    state.errors && setServerActionErrors(state.errors, setError);
  }, [state.errors, setError]);

  useEffect(() => {
    !isEmpty(taskValue) && reset(taskFormInitValues);
  }, [taskFormInitValues, taskValue, reset]);

  return (
    <form className="dark:text-white" action={handleSubmit}>
      <TaskFormContent
        assginedToOptions={assginedToOptions}
        listProject={fromProject}
        control={control}
        setValue={setValue}
        responseMessage={state?.error}
        isDisabled={isDisabled}
        isCreated={isEmpty(taskValue)}
      />
    </form>
  );
};
