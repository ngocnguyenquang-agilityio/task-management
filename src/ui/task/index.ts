export { TaskForm } from './TaskForm';
export { TaskTable } from './TaskTable';
export { CreateTaskFormWrapper } from './CreateTaskFormWrapper';
export { EditTaskFormWrapper } from './EditTaskFormWrapper';
export { EditTaskContainer } from './EditTaskContainer';
export { FilterWrapper } from './FilterWrapper';
