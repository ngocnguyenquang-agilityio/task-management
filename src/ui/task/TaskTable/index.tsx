import { Suspense } from 'react';
import Link from 'next/link';

// Auths
import { auth } from '@/auth';

// APIs
import { getProjectList, getTasks } from '@/api';

// Components
import {
  ErrorMessage,
  ItemNotFound,
  PaginationWrapper,
  Text,
} from '@/components';
import { FilterWrapper } from '../FilterWrapper';

// Constants
import {
  DATE_FORMAT,
  ROUTES,
  TASK_PRIORITY_VALUE,
  TASK_STATUS_VALUE,
  FIELDS,
  LIMIT_ITEMS,
  ORDER_TYPES,
  QUERY_PARAMS,
} from '@/constants';

// Icons
import { MdArrowRightAlt } from 'react-icons/md';

// Types
import { QueryFilter, SearchParams, VariantType } from '@/types';

// Utils
import { formatDate } from '@/utils';

interface TaskTableProps {
  searchParams: SearchParams;
}

const labelMapping = {
  [TASK_STATUS_VALUE.NOT_STARTED]: {
    value: 'Not Started',
    variant: 'primary',
  },
  [TASK_STATUS_VALUE.IN_PROGRESS]: {
    value: 'In Progress',
    variant: 'secondary',
  },
  [TASK_STATUS_VALUE.DONE]: {
    value: 'Done',
    variant: 'success',
  },
  [TASK_PRIORITY_VALUE.LOW]: {
    value: 'Low',
    variant: 'quaternary',
  },
  [TASK_PRIORITY_VALUE.MEDIUM]: {
    value: 'Medium',
    variant: 'active',
  },
  [TASK_PRIORITY_VALUE.HIGH]: {
    value: 'High',
    variant: 'error',
  },
};

export const TaskTable = async ({ searchParams }: TaskTableProps) => {
  const session = await auth();

  const { page, sortBy, status, priority, projectId, filterByUser } =
    searchParams;

  const query: QueryFilter[] = [];

  if (status) {
    const statusFilterList = decodeURIComponent(status)?.split(',');

    query.push({
      field: QUERY_PARAMS.STATUS,
      comparison: 'in',
      value: statusFilterList,
    });
  }

  if (priority) {
    const priorityFilterList = decodeURIComponent(priority)?.split(',');

    query.push({
      field: QUERY_PARAMS.PRIORITY,
      comparison: 'in',
      value: priorityFilterList,
    });
  }

  if (projectId) {
    const projectIdFilterList = decodeURIComponent(projectId)?.split(',');

    query.push({
      field: QUERY_PARAMS.PROJECT_ID,
      comparison: 'in',
      value: projectIdFilterList,
    });
  }

  if (filterByUser && session) {
    query.push({
      field: QUERY_PARAMS.ASSIGNED_TO,
      comparison: '==',
      value: session.user.id,
    });
  }

  const {
    data: taskListData,
    error: taskError,
    total,
  } = await getTasks({
    page: parseInt(page || '1'),
    limitItem: LIMIT_ITEMS.DEFAULT,
    orderItem: {
      field: FIELDS.UPDATED_AT,
      type: sortBy || ORDER_TYPES.DESC,
    },
    query,
  });

  const { data: projectListData, error: projectListError } = await getProjectList({
    orderItem: {
      field: FIELDS.UPDATED_AT,
      type: ORDER_TYPES.DESC,
    },
    query: session
      ? []
      : [
          {
            field: QUERY_PARAMS.IS_PUBLIC,
            comparison: '==',
            value: true,
          },
        ],
  });

  const error = projectListError || taskError;

  if (error) return <ErrorMessage message={error} />;

  return (
    <div className="relative overflow-x-auto">
      <FilterWrapper
        showFilterCheckbox={!!session}
        showPriorityFilter={true}
        showStatusFilter={true}
        projectList={projectListData}
        checkboxLabel="My Task(s)"
      />
      {taskListData.length === 0 ? (
        <ItemNotFound
          title="Empty Tasks"
          description="Your task are currently empty. Please create new task or stay tuned for updates"
        />
      ) : (
        <table className="w-full text-sm text-left text-gray-500">
          <thead className="text-xs uppercase text-gray-700 bg-gray-50 dark:bg-zinc-700 dark:text-gray-300">
            <tr className="font-bold">
              <th className="px-6 py-4">Task Name</th>
              <th className="px-2">Status</th>
              <th className="px-2">Date</th>
              <th className="px-2">Priority</th>
              <th className="px-2 hidden sm:block mt-3.5">Estimation</th>
              <th className="w-6" />
            </tr>
          </thead>
          <tbody className="bg-white dark:bg-zinc-800">
            {taskListData.map((task) => {
              const { value: statusValue, variant: statusVariant } =
                labelMapping[task.status];
              const { value: priorityValue, variant: priorityVariant } =
                labelMapping[task.priority];

              return (
                <tr
                  key={task.id}
                  className="border-b-2 px-8 py-5 rounded-lg hover:bg-zinc-300 dark:hover:bg-gray-700 text-sm"
                >
                  <td className="px-6 py-4 font-medium text-gray-900 dark:text-white whitespace-nowrap w-full max-w-0 xl:w-auto">
                    <Link
                      href={
                        session
                          ? ROUTES.ADMIN_TASK_DETAIL(task.id)
                          : ROUTES.TASK_DETAIL(encodeURIComponent(task.slug))
                      }
                    >
                      <p className="truncate">{task.title}</p>
                    </Link>
                  </td>
                  <td className="min-w-20 sm:min-w-28 px-2 py-4">
                    <Text
                      value={statusValue}
                      variant={statusVariant as VariantType}
                      customClass="font-medium px-0"
                    />
                  </td>
                  <td>
                    <p className="text-gray-900 dark:text-white min-w-16 sm:min-w-28 pl-2">
                      {formatDate(task.createdAt, DATE_FORMAT.Secondary)}
                    </p>
                  </td>
                  <td className="px-2 py-4">
                    <Text
                      value={priorityValue}
                      variant={priorityVariant as VariantType}
                      customClass="font-medium px-0"
                    />
                  </td>
                  <td className="hidden sm:block mt-3.5">
                    <Text
                      customClass="w-full text-gray-900 dark:text-white"
                      value="1 hour"
                    />
                  </td>
                  <td>
                    <Link
                      href={
                        session
                          ? ROUTES.ADMIN_TASK_DETAIL(task.id)
                          : ROUTES.TASK_DETAIL(encodeURIComponent(task.slug))
                      }
                      aria-label="task-detail"
                      className="px-2 md:px-4 flex justify-end"
                    >
                      <MdArrowRightAlt className="w-6 h-6 text-gray-900 dark:text-white" />
                    </Link>
                  </td>
                </tr>
              );
            })}
          </tbody>
        </table>
      )}

      <div className="mt-10 sm:mt-6 flex justify-end">
        {total && (
          <Suspense>
            <PaginationWrapper total={total} pageSize={LIMIT_ITEMS.DEFAULT} />
          </Suspense>
        )}
      </div>
    </div>
  );
};
