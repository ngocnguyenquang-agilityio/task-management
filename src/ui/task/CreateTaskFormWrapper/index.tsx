'use client';

import { useFormState } from 'react-dom';

// APIs
import { addTaskToProject } from '@/actions';

// Components
import { TaskForm } from '../TaskForm';

// Models
import { Participation, Project } from '@/models';

type CreateTaskFormWrapperProps = {
  memberOptions: Participation[];
  project: Project;
};

export const CreateTaskFormWrapper = ({
  memberOptions,
  project,
}: CreateTaskFormWrapperProps) => {
  const initialState = { message: null, formErrors: {} };
  const [state, dispatch] = useFormState(addTaskToProject, initialState);

  return (
    <TaskForm
      assginedToOptions={memberOptions}
      fromProject={project}
      state={state}
      onSubmit={dispatch}
    />
  );
};
