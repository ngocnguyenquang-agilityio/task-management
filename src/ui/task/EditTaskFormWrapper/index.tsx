'use client';

import { useFormState } from 'react-dom';

// APIs
import { editTask } from '@/actions';

// Components
import { TaskForm } from '../TaskForm';

// Models
import { User } from '@/types';
import { Project, Task } from '@/models';

type CreateTaskFormWrapperProps = {
  memberOptions: User[];
  listProject: Project[];
  taskData: Task;
};

export const EditTaskFormWrapper = ({
  memberOptions,
  listProject,
  taskData,
}: CreateTaskFormWrapperProps) => {
  const initialState = { message: null, formErrors: {} };
  const editTaskWithId = editTask.bind(null, taskData.id);
  const [state, dispatch] = useFormState(editTaskWithId, initialState);

  return (
    <TaskForm
      state={state}
      assginedToOptions={memberOptions}
      fromProject={listProject}
      taskValue={taskData}
      onSubmit={dispatch}
    />
  );
};
