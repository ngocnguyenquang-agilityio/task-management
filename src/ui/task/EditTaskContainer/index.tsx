// APIs
import { getPartipationsByProjectId, getProjectById } from '@/api';

// Components
import { ErrorMessage, ItemNotFound } from '@/components';
import { EditTaskFormWrapper } from '../EditTaskFormWrapper';

// Constants
import { ERROR_MESSAGES } from '@/constants';

// Models
import { Task } from '@/models';

type EditTaskContainerProps = {
  taskData: Task;
  isReadOnly?: boolean;
};

export const EditTaskContainer = async ({
  taskData,
  isReadOnly = false,
}: EditTaskContainerProps) => {
  const projectId = taskData.projectId;
  const { data: participantList, error: participantListError } =
    await getPartipationsByProjectId(projectId);
  const { data: projectData, error: projectError } =
    await getProjectById(projectId);

  const error = participantListError || projectError;

  if (error) {
    return <ErrorMessage message={error} />;
  }
  if (!projectData) {
    return (
      <ItemNotFound
        title={ERROR_MESSAGES.DATA_NOT_FOUND}
        description="Please create new user or new project to proceed."
        customClass={{
          wrapper: 'rounded-lg px-5 bg-white py-20',
        }}
      />
    );
  }
  return (
    <EditTaskFormWrapper
      memberOptions={participantList}
      project={projectData}
      taskData={taskData}
      isReadOnly={projectData.isArchived || isReadOnly}
    />
  );
};
