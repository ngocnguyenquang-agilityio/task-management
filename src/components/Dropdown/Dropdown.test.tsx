import { render, screen, fireEvent } from '@testing-library/react';

// Component
import { Dropdown } from '.';

import { SORT_OPTIONS } from '@/ui/task/FilterWrapper';

const mockProps = {
  options: SORT_OPTIONS,
  onSelect: jest.fn(),
};

describe('Dropdown component', () => {
  test('Render component correctly', () => {
    expect(render(<Dropdown {...mockProps} />)).toMatchSnapshot();
  });

  test('Show the selected item', () => {
    const { getByText } = render(
      <Dropdown {...mockProps} selectedItemValue="desc" />,
    );
    const comp = getByText('Desc');
    expect(comp).toBeInTheDocument();
  });

  test('Show list options', async () => {
    render(<Dropdown {...mockProps} />);
    const button = screen.getByRole('button');
    await fireEvent.click(button);
    const options = screen.getByTestId('options');
    expect(options).toBeInTheDocument();
  });

  test('Should render correct number of options', async () => {
    render(<Dropdown {...mockProps} />);
    const button = screen.getByRole('button');
    expect(button).toBeInTheDocument();
    await fireEvent.click(button);
    const options = screen.getByTestId('options');
    expect(options.children).toHaveLength(SORT_OPTIONS.length);
  });

  test('Should call onSelect function', async () => {
    render(<Dropdown {...mockProps} />);
    const button = screen.getByRole('button');
    await fireEvent.click(button);
    await fireEvent.click(screen.getByText('Desc'));
    expect(mockProps.onSelect).toHaveBeenCalledTimes(1);
  });
});
