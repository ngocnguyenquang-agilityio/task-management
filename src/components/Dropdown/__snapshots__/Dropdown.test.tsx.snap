// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown component Render component correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        tabindex="1"
      >
        <button
          class="rounded-lg disabled:cursor-not-allowed border border-gray-200 hover:bg-gray-100 dark:border-white w-full p-5 bg-zinc-50 dark:text-white dark:bg-neutral-900 dark:hover:bg-neutral-800 dark:hover:text-white flex justify-between items-center gap-2 outline outline-1 outline-zinc-300 text-zinc-400"
          type="button"
        >
          <span>
            Dropdown
          </span>
          <div
            class=""
          >
            <svg
              class="h-6 w-6 text-slate-500 transform duration-300 ease-in-out"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M5.22 8.22a.749.749 0 0 0 0 1.06l6.25 6.25a.749.749 0 0 0 1.06 0l6.25-6.25a.749.749 0 1 0-1.06-1.06L12 13.939 6.28 8.22a.749.749 0 0 0-1.06 0Z"
              />
            </svg>
          </div>
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      tabindex="1"
    >
      <button
        class="rounded-lg disabled:cursor-not-allowed border border-gray-200 hover:bg-gray-100 dark:border-white w-full p-5 bg-zinc-50 dark:text-white dark:bg-neutral-900 dark:hover:bg-neutral-800 dark:hover:text-white flex justify-between items-center gap-2 outline outline-1 outline-zinc-300 text-zinc-400"
        type="button"
      >
        <span>
          Dropdown
        </span>
        <div
          class=""
        >
          <svg
            class="h-6 w-6 text-slate-500 transform duration-300 ease-in-out"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5.22 8.22a.749.749 0 0 0 0 1.06l6.25 6.25a.749.749 0 0 0 1.06 0l6.25-6.25a.749.749 0 1 0-1.06-1.06L12 13.939 6.28 8.22a.749.749 0 0 0-1.06 0Z"
            />
          </svg>
        </div>
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
